docs: init user docs (#2423)
GreptimeTeam/docs: docs/enterprise/user.md

---
keywords: [User, Permission, GreptimeDB Enterprise, RBAC, ACL, Authentication]
description: The overview of GreptimeDB User and Permission mechanism.
---

# Built-in User Management

GreptimeDB Enterprise provides a built-in user and permission system backed by

> the Meta Server. It supports Role-Based Access Control (RBAC) and fine-grained

Reviewer: It's a term:
```suggestion
the Metasrv. It supports Role-Based Access Control (RBAC) and fine-grained
```

Access Control Lists (ACLs) to ensure data security and isolation.

## Key Features

- **Built-in User Management**: User accounts and permissions are stored in the

>   Meta Server, ensuring consistent management across the cluster.

Reviewer: Please replace all the Meta server into `Metasrv`
```suggestion
  Metasrv, ensuring consistent management across the cluster.
```

- **Role-Based Access Control (RBAC)**: Assign global privileges to users,
  controlling operations like `SELECT`, `INSERT`, `CREATE TABLE`, and more.

> - **Fine-grained ACLs**: Control table-level access within specific schemas

Reviewer: Prefer `database`:
```suggestion
- **Fine-grained ACLs**: Control table-level access within specific databases
```

  using exact matches or regular expressions.

> - **Dynamic Management**: Manage users dynamically via HTTP APIs without

Reviewer: Does the enterprise dashboard  web UI support management?

Author: Yes. The newer version of enterprise dashboard comes with the user management.
But it requires some configuration to enable the feature.

Should we only demonstrate how to use the user management UI in this documentation instead of these API details?

```

> **Note**: The trailing colon `:` is required by the configuration parser.

### 2. Initial Account Seeding (Optional)

The password file uses the following format:
`<username>[:<role>]=<password>`

Available roles:
- `admin`: Full privileges, including user management.
- `readonly` (or `ro`): Read-only access (`SqlSelect`).
- `writeonly` (or `wo`): Write-only access (e.g., `SqlInsert`, `TableCreate`).
- `readwrite` (or `rw`): Both read and write access (default).

Example `passwords.txt`:

```text
# username[:role]=password
superuser:admin=strong_password
alice:ro=alice_pwd
bob:rw=bob_pwd
```

Seeded users are granted full access (`AclType::All`) to the `public` schema by default.

### 3. Manage Users via HTTP API

Once the server is running, admin users can manage other users via the `/v1/users` HTTP API.

#### List all users

```bash
curl -u admin:strong_password http://localhost:4000/v1/users
```

Example response:

```json
[
  {
    "username": "admin",
    "catalog": "greptime",
    "privileges": 131071,
    "acl_map": {
      "public": [{ "type": "all" }]
    },
    "created_at": 1704067200000,
    "updated_at": 1704067200000
  }
]

> ```

Reviewer: Describe the keys when they first appear.

#### Get a specific user

```bash
curl -u admin:strong_password http://localhost:4000/v1/users/charlie
```

> **Note**: Non-admin users can only retrieve their own user information.

#### Get current user

```bash
curl -u charlie:charlie_password http://localhost:4000/v1/users/current
```

> 

Reviewer: Better to add the sample response.

This endpoint returns the authenticated user's information without requiring
admin privileges.

#### Create a new user

```bash
curl -u admin:strong_password -X POST http://localhost:4000/v1/users \
  -H "Content-Type: application/json" \
  -d '{
    "username": "charlie",
    "password": "charlie_password",
    "privileges": 8,

>     "acl_map": {

Reviewer: Explain the functionality and configuration of the `acl_map`.

      "public": [
        { "type": "match", "table": "cpu" },
        { "type": "regex", "pattern": "mem_.*" }
      ]
    }
  }'
```

> **Note**: `privileges` is a bitmask of `UserPrivilege`. `8` corresponds to `SqlSelect`.

#### Update a user

```bash
curl -u admin:strong_password -X PUT http://localhost:4000/v1/users/charlie \
  -H "Content-Type: application/json" \
  -d '{
    "password": "new_secure_password",
    "privileges": 24
  }'
```

At least one field (`password`, `privileges`, or `acl_map`) must be provided.

#### Delete a user

```bash
curl -u admin:strong_password -X DELETE http://localhost:4000/v1/users/charlie
```

> **Note**: The `admin` user cannot be deleted.

## Privileges and ACLs

### Global Privileges

> Privileges are defined as a bitmask of the following operations:

Reviewer: Bitmasks are unnecessary for users.

---
keywords: [User, Permission, GreptimeDB Enterprise, RBAC, ACL, Authentication]
description: The overview of GreptimeDB User and Permission mechanism.
---

# User and Permission System

GreptimeDB Enterprise provides a persistent user and permission system based on
a persistent metadata backend (Meta Server). It supports Role-Based Access
Control (RBAC) and fine-grained Access Control Lists (ACLs) to ensure data
security and isolation.

## Key Features

> - **Persistent Management**: User accounts and permissions are stored in the

Reviewer 2: Built-in User Management
